services/typeddata/hash_test.go

```diff
@@ -0,0 +1,261 @@
+func TestTypeString(t *testing.T) {
+	type testCase struct {
+		description string
+		typeString  string
+		types       Types
+		target      string
+	}
+	for _, tc := range []testCase{
+		{
+			"WithoutDeps",
+			"Person(string name,address wallet)",
+			Types{"Person": []Field{{Name: "name", Type: "string"}, {Name: "wallet", Type: "address"}}},
+			"Person",
+		},
+		{
+			"SingleDep",
+			"Mail(Person from,Person to)Person(string name,address wallet)",
+			Types{
+				"Person": []Field{{Name: "name", Type: "string"}, {Name: "wallet", Type: "address"}},
+				"Mail":   []Field{{Name: "from", Type: "Person"}, {Name: "to", Type: "Person"}},
+			},
+			"Mail",
+		},
+		{
+			"DepsOrdered",
+			"Z(A a,B b)A(string name)B(string name)",
+			Types{
+				"A": []Field{{Name: "name", Type: "string"}},
+				"B": []Field{{Name: "name", Type: "string"}},
+				"Z": []Field{{Name: "a", Type: "A"}, {Name: "b", Type: "B"}},
+			},
+			"Z",
+		},
+		{
+			"RecursiveDepsIgnored",
+			"Z(A a)A(Z z)",
+			Types{
+				"A": []Field{{Name: "z", Type: "Z"}},
+				"Z": []Field{{Name: "a", Type: "A"}},
+			},
+			"Z",
+		},
+	} {
+		tc := tc
+		t.Run(tc.description, func(t *testing.T) {
+			require.Equal(t, tc.typeString, typeString(tc.target, tc.types))
+		})
+	}
+}
+
+func TestEncodeData(t *testing.T) {
+	type testCase struct {
+		description string
+		message     map[string]interface{}
+		types       Types
+		target      string
+		result      func(testCase) common.Hash
+	}
+
+	bytes32, _ := abi.NewType("bytes32")
+	addr, _ := abi.NewType("address")
+	bool, _ := abi.NewType("bool")
+
+	for _, tc := range []testCase{
+		{
+			"HexAddressConvertedToBytes",
+			map[string]interface{}{"wallet": "0xCD2a3d9F938E13CD947Ec05AbC7FE734Df8DD826"},
+			Types{"A": []Field{{Name: "wallet", Type: "address"}}},
+			"A",
+			func(tc testCase) common.Hash {
+				args := abi.Arguments{{Type: bytes32}, {Type: addr}}
+				typehash := typeHash(tc.target, tc.types)
+				packed, _ := args.Pack(typehash,
+					common.HexToAddress(tc.message["wallet"].(string)))
+				return crypto.Keccak256Hash(packed)
+			},
+		},
+		{
+			"StringHashed",
+			map[string]interface{}{"name": "AAA"},
+			Types{"A": []Field{{Name: "name", Type: "string"}}},
+			"A",
+			func(tc testCase) common.Hash {
+				args := abi.Arguments{{Type: bytes32}, {Type: bytes32}}
+				typehash := typeHash(tc.target, tc.types)
+				packed, _ := args.Pack(typehash,
+					crypto.Keccak256Hash([]byte(tc.message["name"].(string))))
+				return crypto.Keccak256Hash(packed)
+			},
+		},
+		{
+			"BytesHashed",
+			map[string]interface{}{"name": []byte{1, 2, 3}},
+			Types{"A": []Field{{Name: "name", Type: "bytes"}}},
+			"A",
+			func(tc testCase) common.Hash {
+				args := abi.Arguments{{Type: bytes32}, {Type: bytes32}}
+				typehash := typeHash(tc.target, tc.types)
+				packed, _ := args.Pack(typehash,
+					crypto.Keccak256Hash(tc.message["name"].([]byte)))
+				return crypto.Keccak256Hash(packed)
+			},
+		},
+		{
+			"FixedBytesAsIs",
```

Comment:
Why fixed-length slice is treated differently? Does `abi` check the length to be 32 bytes?